src/zino/planned_maintenance.py

```diff
@@ -0,0 +1,217 @@
+        type: Literal["portstate", "device"],
+        match_type: Literal["regexp", "str", "exact", "intf-regexp"],
+        match_expression: str,
+        match_device: Optional[str],
+    ) -> PlannedMaintenance:
+        """Creates a planned maintenance, adds it to the planned_maintenances dict and
+        returns it
+        """
+        pm_id = self.get_next_available_pm_id()
+        pm = PlannedMaintenance(
+            id=pm_id,
+            start_time=start_time,
+            end_time=end_time,
+            type=type,
+            match_type=match_type,
+            match_expression=match_expression,
+            match_device=match_device,
+        )
+        self.planned_maintenances[pm_id] = pm
+        self._call_observers()
+        return pm
+
+    def close_planned_maintenance(self, id: int, reason: str, user: str) -> None:
+        """Deletes planned maintenance with the given id"""
+        # See `close` function in Zino 1.0 `pm.tcl`
+        pm = self.planned_maintenances.get(id, None)
+        if not pm:
+            # TODO figure out if this is enough
+            return
+        pm.add_log(f"PM closed by {user}: {reason}")
+        del self.planned_maintenances[id]
+        self._call_observers()
+
+    def get_started_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that have began since the last run of this
+        task until `now`
+        """
+        if self.last_run:
+            return [
+                pm for pm in self.planned_maintenances.values() if self.last_run < pm.start_time <= now < pm.end_time
+            ]
+        else:
+            return [pm for pm in self.planned_maintenances.values() if pm.start_time <= now < pm.end_time]
+
+    def get_ended_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that have ended since the last run of this
+        task until `now`
+        """
+        if self.last_run:
+            return [pm for pm in self.planned_maintenances.values() if self.last_run < pm.end_time <= now]
+        else:
+            return [pm for pm in self.planned_maintenances.values() if pm.end_time <= now]
+
+    def get_active_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that are currently active
+
+        This means it has started before `now` and the end_time is later than `now`
+        """
+        return [pm for pm in self.planned_maintenances.values() if pm.start_time < now < pm.end_time]
+
+    def get_old_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that should get deleted
+
+        This means that `now` is 3 days later than end_time
+        """
+        return [pm for pm in self.planned_maintenances.values() if now - pm.end_time > datetime.timedelta(days=3)]
```

Comment:
Magic number alert.  Not sure if the number is configurable in Zino 1, but the timedelta should be defined as a constant somewhere, e.g.

```python
PM_EXPIRY_TIME = datetime.timedelta(days=3)
```

Reply:
It is not configurable in Zino 1, it is directly defined in the code (see `old_pms` in `pm.tcl`)